### PR TITLE
[exporter/prometheusexporter] accumulate delta temporality histograms

### PR DESCRIPTION
**Pull request by hkfgo**
**Description:** <Describe what has changed.>
This continues the work done in the now closed [PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/20530). I have addressed issues raised in the original PR by
- Adding logic to handle timestamp misalignments 
- Adding fix + a out-of-bounds bug  

In addition, I have performed end-to-end testing in a local setup, and confirmed that accumulated histogram time series are correct. 

**Link to tracking Issue:** <Issue number if applicable>
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/4968
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/9006
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19153
**Testing:** <Describe what testing was performed and which tests were added.>
Added tests for timestamp misalignment and an out-of-bounds bug discovered in the previous PR. 
End-to-end testing to ensure histogram bucket counts exported to Prometheus are correct

### CODE REVIEW

File: exporter/prometheusexporter/accumulator.go
Comment (bitomaxsp) [inline exporter/prometheusexporter/accumulator.go:244]: typo: idenity

Comment (hkfgo) [reply exporter/prometheusexporter/accumulator.go:244]: Oh oops, some of the comments were actually just for my own understanding of the code. Let me go ahead and remove those.

File: exporter/prometheusexporter/accumulator_test.go
Comment (Aneurysm9) [inline exporter/prometheusexporter/accumulator_test.go:65]: Why are these out of order and where is `m1`?

File: exporter/prometheusexporter/accumulator_test.go
Comment (Aneurysm9) [inline exporter/prometheusexporter/accumulator_test.go:96]: Are these independent tests or does one test affect the operation of the next?  Can they be separated out?  It's somewhat hard to follow what's being done here.

Comment (hkfgo) [reply exporter/prometheusexporter/accumulator_test.go]: Agreed that having all the test setup in a single place makes it hard to follow. These were not independent tests, but I just did a re-write to have independent setups for each. Hopefully the tests are easier to follow now.

File: exporter/prometheusexporter/accumulator_test.go
Comment (hkfgo) [reply exporter/prometheusexporter/accumulator_test.go]: The numbers were meant to be ordered chronologically I believe(I didn't write most of the tests). I re-wrote all of the tests just now to avoid these confusions; marking this as resolved

File: exporter/prometheusexporter/accumulator.go
Comment (hkfgo) [reply exporter/prometheusexporter/accumulator.go]: I added this check for disagreement 
```
else if ip.HasSum() != pp.HasSum() {
    a.logger.With(
	    zap.String("metric_name", metric.Name()),
    ).Warn("Dropped histogram data point due to disagreement on sum tracking")
    continue
}
``` 
However, I could use more of your help here, as I am not sure if I'm doing it right(I also didn't write the original block of code). 

For one, when does it make for a histogram metric to come without sum? In other words, can we simply assume they have sums always? 

Also, assuming it makes sense for a sum to be absent(and thus the necessity to check), should we also be checking for presence of a min and max then?

File: exporter/prometheusexporter/accumulator.go
Comment (hkfgo) [reply exporter/prometheusexporter/accumulator.go]: Committed this suggestion

File: exporter/prometheusexporter/accumulator.go
Comment (hkfgo) [reply exporter/prometheusexporter/accumulator.go]: Speaking of min and max, I just realized the code as written right now will not record any min and max values. That would be a bug correct?

Comment (Aneurysm9) [reply exporter/prometheusexporter/accumulator.go]: The Prometheus metrics data model does not include min/max for histograms as all components of a Prometheus histogram are actually counter series and counters are cumulative monotonic sums.  It wouldn't be possible to track a lower `min` value once the first minimum was set and it doesn't generally make sense to track min/max over the entire lifetime of a series.  Instead it makes more sense to be able to track the min/max observed during any given window, such as is possible with a delta histogram.

From the [OTel metrics spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#histogram):

> The aggregation temporality also has implications on the min and max fields. Min and max are more useful for Delta temporality, since the values represented by Cumulative min and max will stabilize as more events are recorded. Additionally, it is possible to convert min and max from Delta to Cumulative, but not from Cumulative to Delta. When converting from Cumulative to Delta, min and max can be dropped, or captured in an alternative representation such as a gauge.

Comment (hkfgo) [reply exporter/prometheusexporter/accumulator.go]: Got it. What do you think about that `HasSum` check? The OTEL spec you linked states that sum is actually a required field. In that case, I would think it'd be safe to remove the check and make the code look much nicer

Comment (hkfgo) [reply exporter/prometheusexporter/accumulator.go]: Hey Anthony, I went ahead and remove the `HasSum` check. The reason was what I stated above, plus the fact that, base on my research, we can generally assume good data points in exporter as receiver should have handled many of the bad cases. Let me know if that makes sense. Thanks!

**Comment by Aaron-ML** [comment]
@hkfgo is this still being worked on? patiently waiting to get some of this functionality for our use case. Thanks!

**Review by AlexDCraig — APPROVED** [review]

**Review by AlexDCraig — COMMENTED** [review]
Also interested in this functionality

**Comment by hkfgo** [comment]
> @hkfgo is this still being worked on? patiently waiting to get some of this functionality for our use case. Thanks!

Hey @Aaron-ML I'm also patiently waiting for a code owner to approve this 😢. I understand that code owners are all very busy though, so no hard feelings. 

@Aneurysm9 Would you mind taking a look again? Thanks so much!

**Comment by hkfgo** [comment]
I am doing some "integration testing" in a local docker-compose setup while waiting for more feedback. The set up looks like (OpenTelemetry SDK -> OTEL Collector Build -> Prometheus). Will report back on testing results

**Comment by hkfgo** [comment]
For end-to-end testing, I had a thread continuously recording data points in a loop like this 
`    histogramExplicit.record(100);
      histogramExplicit.record(500);
      histogramExplicit.record(2500);
      histogramExplicit.record(5000);`

The data points are then exported to otel-collector and Prometheus. The testing did discover a bug - I incorrectly assumed start timestamp of delta histogram is static, when it in fact is constantly moving for every export interval. After fixing that, the resulting bucket counts look good :) See the attached screenshot. 

Screenshot 2023-08-29 at 12 01 28 PM

**Comment by ggreen53** [comment]
Would like to add my enthusiasm to see this PR land as it will solve a number of our issues.

**Comment by djaglowski** [comment]
@Aneurysm9 or perhaps @dashpole, could you take a look?

**Comment by bitomaxsp** [comment]
Any help needed with this?

**Comment by hkfgo** [comment]
> Any help needed with this?

@bitomaxsp I think the PR is in a good shape personally. I've built and tested it out, and it worked fine. The number one thing is to get a review from a maintainer/approver. 

You are more than welcome to review this PR though. I see two test failures too. I believe those are log receiver/processor tests, so should be unrelated to my changes. Please feel free to advise on that too. I'm not familiar with these end-to-end tests.

**Comment by mageddo** [comment]
Is there a roadmap to that feature? It seems a lot of people are waiting for it.

**Comment by hkfgo** [comment]
> Is there a roadmap to that feature? It seems a lot of people are waiting for it.
IMO this PR just really needs a maintainer review again. From my point of view, this PR has passed my end-to-end testing and there's no further action I can think of right now. If the reviewer has feedback then I'm happy to address them after, ofc. 

Perhaps people waiting on this feature can help out and get this PR reviewed again?

**Comment by hampusrosvall** [comment]
Also very interested in this! Happy to help out if needed. How can we get attention from code owners?

**Comment by hkfgo** [comment]
> Also very interested in this! Happy to help out if needed. How can we get attention from code owners?

I've been pinging for a review on CNCF Slack.

**Comment by ggreen53** [comment]
I built this change into a my deployed collector and confirmed it works as expected.

**Comment by jpkrohling** [comment]
@sh0rez, can you please review this one?

File: exporter/prometheusexporter/accumulator.go
Comment (atoulme) [inline exporter/prometheusexporter/accumulator.go:221]:
```suggestion
```

File: exporter/prometheusexporter/accumulator.go
Comment (atoulme) [inline exporter/prometheusexporter/accumulator.go:204]: I appreciate this clarity, but can this rename be split out into a separate chore PR so we can concentrate on the changes here?

File: exporter/prometheusexporter/accumulator.go
Comment (atoulme) [inline exporter/prometheusexporter/accumulator.go:242]:
```suggestion
```

File: exporter/prometheusexporter/accumulator.go
Comment (atoulme) [reply exporter/prometheusexporter/accumulator.go:233]: are you still looking to fix this typo?

File: exporter/prometheusexporter/accumulator.go
Comment (atoulme) [inline exporter/prometheusexporter/accumulator.go:259]:
```suggestion
				// treat misalignment as restart and reset, or violation of single-writer principle and drop
```

File: exporter/prometheusexporter/accumulator.go
Comment (atoulme) [inline exporter/prometheusexporter/accumulator.go:365]: if match is used in the for loop condition, can we move this inside the `if match` block below?

Comment (atoulme) [reply exporter/prometheusexporter/accumulator.go:365]: ok never mind. I don't understand what this does.

**Comment by hkfgo** [comment]
Thanks for the review @atoulme! I've tidied-up the code a bit following your comments. Would you mind pinging other maintainers to have a look as well?

**Comment by jpkrohling** [comment]
> @jpkrohling @sh0rez are you still looking to review?

Sorry about that, I think it's not realistic to wait for a review from us to move forward with this PR.

**Comment by hkfgo** [comment]
> > @jpkrohling @sh0rez are you still looking to review?
> 
> Sorry about that, I think it's not realistic to wait for a review from us to move forward with this PR.

Ack. @atoulme would you be comfortable giving this an approval? It's been thoroughly tested and follows an existing pattern in Prometheus exporter to accumulate delta counters.

**Review by atoulme — APPROVED** [review]

**Comment by gburton1** [comment]
This is approved, and I think it just needs to be merged by an authorized user now. Can someone do that?

**Pull request merged**
